quality_of_service_demo/src/lifespan.cpp

```diff
@@ -0,0 +1,115 @@
+static const size_t DEFAULT_SUBSCRIBE_AFTER = 5000;
+
+void print_usage()
+{
+  printf("Usage for lifespan:\n");
+  printf("lifespan lifespan_duration [--history history_depth] [-p publish_n_messages] "
+    "[-s subscribe_after_duration] [-h]\n");
+  printf("required arguments:\n");
+  printf("lifespan duration: Duration (in ms) of the Lifespan QoS setting.\n");
+  printf("options:\n");
+  printf("-h : Print this help message.\n");
+  printf("%s history : The depth of the Publisher's history queue - "
+    "the maximum number of messages it will store for late-joining subscriptions. "
+    "Defaults to %zu\n",
+    OPTION_HISTORY, DEFAULT_HISTORY);
+  printf("%s publish_n_messages : How many messages to publish before stopping. "
+    "Defaults to %zu\n",
+    OPTION_PUBLISH_COUNT, DEFAULT_PUBLISH_COUNT);
+  printf("%s subscribe_after_duration : The Subscriber will be created this long (in ms) after "
+    "application startup. Defaults to %zu\n",
+    OPTION_SUBSCRIBE_AFTER, DEFAULT_SUBSCRIBE_AFTER);
+}
+
+int main(int argc, char * argv[])
+{
+  // Force flush of the stdout buffer.
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  // Argument parsing
+  if (argc < 2 || rcutils_cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  // Required arguments
+  std::chrono::milliseconds lifespan_duration(std::stoul(argv[1]));
+
+  // Optional argument default values
+  size_t history = DEFAULT_HISTORY;
+  size_t publish_count = DEFAULT_PUBLISH_COUNT;
+  std::chrono::milliseconds subscribe_after_duration(DEFAULT_SUBSCRIBE_AFTER);
+
+  // Optional argument parsing
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_HISTORY)) {
+    history = std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_HISTORY));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_PUBLISH_COUNT)) {
+    publish_count = std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_PUBLISH_COUNT));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_SUBSCRIBE_AFTER)) {
+    subscribe_after_duration = std::chrono::milliseconds(
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_SUBSCRIBE_AFTER)));
+  }
+
+  // Configuration and Initialization
+  rclcpp::init(argc, argv);
+  rclcpp::executors::SingleThreadedExecutor exec;
+
+  std::string topic("qos_lifespan_chatter");
+
+  rclcpp::QoS qos_profile(history);
+  qos_profile.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
+  qos_profile.lifespan(lifespan_duration);
+
+  rclcpp::SubscriptionOptions sub_options;
+  rclcpp::PublisherOptions pub_options;
+
+  auto listener = std::make_shared<Listener>(topic, qos_profile, sub_options, true);
+  auto talker = std::make_shared<Talker>(topic, qos_profile, pub_options, publish_count);
+
+  exec.add_node(talker);
+  exec.add_node(listener);
+  auto timer = listener->create_wall_timer(
+    subscribe_after_duration,
+    [listener]() -> void {
+      listener->start_listening();
+    });
```

Comment:
nitpick: in the deadline example you create the timer before adding the nodes to the executor, which I think makes more sense from a flow perspective (the order doesn't really make a difference), but at least we should be consistent otherwise the user might wonder why the examples differ (in vain because they don't).

Reply:
Ok